Updates to LLVM 3.7 and fixes several GPU issues

gkanwar opened the pull request:
This is a cherry-picked remake of PR #39, to avoid the rebase artifacts.

The updated backend uses the approach taken by Halide: an NVPTX backend with LLVM passes to emit a PTX string, instead of calling out to NVVM. This should allow us to keep pace with updating LLVM versions.

There are still two test suites that fail for specific reasons:
-- ffi: We don't have GPU ffi support
-- strings: String intrinsics are not implemented for the GPU yet

We also still do not generate reasonable GPU code for matrix multiplication, since the emitted IR is designed for serial execution.

fredrikbk commented:
This branch does not currently build with llvm (3.7.0 or 3.8.1) built without RTTI. So someone has to build LLVM with RTTI enabled (`-DLLVM_ENABLE_RTTI=ON` on the command line to llvm cmake). This is not great, since RTTI is disabled by default when LLVM builds. I don't see know RTTI causes problems though. Does anyone have any ideas? Can you reproduce this error?
```
Undefined symbols for architecture x86_64:
  "typeinfo for llvm::Instruction", referenced from:
      typeinfo for llvm::GetElementPtrInst in llvm_backend.cpp.o
      typeinfo for llvm::PHINode in llvm_backend.cpp.o
      typeinfo for llvm::InsertValueInst in llvm_codegen.cpp.o
      typeinfo for llvm::GetElementPtrInst in llvm_data_layouts.cpp.o
  "typeinfo for llvm::UnaryInstruction", referenced from:
      typeinfo for llvm::ExtractValueInst in llvm_backend.cpp.o
      typeinfo for llvm::ExtractValueInst in llvm_codegen.cpp.o
      typeinfo for llvm::ExtractValueInst in llvm_data_layouts.cpp.o
  "typeinfo for llvm::CmpInst", referenced from:
      typeinfo for llvm::FCmpInst in llvm_backend.cpp.o
      typeinfo for llvm::ICmpInst in llvm_backend.cpp.o
ld: symbol(s) not found for architecture x86_64
```

Without a fix for this it's going to be rather inconvenient for users who already have llvm installed, but that must rebuild it without RTTI. It may have to do with the way LLVM is included in the Simit library. Another solution is to build Simit with RTTI disabled by default. This requires us to roll our own dynamic cast system and to use exceptions for errors only if it is compiled with RTTI (Halide does this).

gkanwar commented:
@fredrikbk -- I have this issue as well, and have been building LLVM with RTTI enabled. I haven't investigated what the dependency is here, and if it's not intrinsically tied to using the IRBuilder in LLVM 3.7 it would be great if we could remove that dependency!

gkanwar commented on README.md:
```diff
@@ -2,27 +2,28 @@ Simit is a new programming language designed to make it easy to write
+    cd llvm3.7
+    mkdir build
+    cd build
+    cmake -DLLVM_ENABLE_TERMINFO=OFF -DLLVM_TARGETS_TO_BUILD="X86;NVPTX" -DLLVM_ENABLE_ASSERTIONS=ON -DCMAKE_BUILD_TYPE=Release ..
```
Should we add the "-DLLVM_ENABLE_RTTI=ON" flag here until we figure out a way around that?

gkanwar commented:
@fredrikbk - I believe I fixed the RTTI dependence in ebbdb7b. It seems like some set of the IRBuilder functions, if referenced from an RTTI-compiled object triggered including unresolved RTTI information for certain instruction-related classes. Still not sure exactly why, but I moved these out into a single non-RTTI source file (llvm_codegen_nortti.cpp).

fredrikbk commented:
Strange but seems like a decent workaround. I'll test it tomorrow. I don't think it makes sense to require an llvm built with rtti. It would make Simit require a custom llvm build.

fredrikbk commented on src/CMakeLists.txt:
```diff
@@ -71,7 +71,7 @@ set_property(TARGET ${PROJECT_NAME} PROPERTY COMPILE_FLAGS ${LLVM_CPPFLAGS})
 execute_process(COMMAND ${LLVM_CONFIG} --includedir OUTPUT_VARIABLE LLVM_INCLUDES OUTPUT_STRIP_TRAILING_WHITESPACE)
 include_directories("${LLVM_INCLUDES}")
 
-set(LLVM_COMPONENTS "${LLVM_COMPONENTS}" core bitwriter x86 ipo)
```
I removed `bitreader asmparser irreader mirparser` in a later commit since Halide doesn't have them. Do we need them for something?

gkanwar replied:
This was needed only in the GPU backend (to parse and load bitcode libraries), so I moved it into the CMakeLists.txt specific to the gpu.

fredrikbk commented:
Jon suggested we verify that the new nvptx backed does not produce much worse code than the old nvcc backend. Did you try to run the `femTet` code from `simit-eval` with the new backend to see if it's still in the same ballpark. Can't see why it wouldn't be, but if it's way of then that would be a big problem.

fredrikbk commented:
I've tested the workaround and it all seems to compile fine. Nice work!

As soon as we've verified the performance we can merge it and send out emails notifying users that we've swapping llvm versions.

gkanwar commented:
Good point on verifying performance. I'll take a look and confirm nothing slipped (it's possible since the backend didn't even build for quite a while).

Re the modules, per my inline comment we need those for the GPU backend to load bitcode libraries. I moved it into that GPU-specific CMakeLists.txt.

jrk commented:
Thanks for doing this, @gkanwar. I'm curious: how does the performance look with NVPTX on the latest LLVM?

gkanwar commented:
@jrk - No problem! I haven't gotten around to running the comparison yet. Unfortunately a couple of other things came up this week, but I should have more info within a day or two :)

gkanwar commented:
Looks like we do in fact take a performance hit from this change. A quick evaluation of isprings on bunny.1 gives 45-50ms per frame vs. 25-35ms per frame on the CPU. Seems like this is worth some more investigation as to the root cause of the performance difference. Since the GPU implementation hasn't been functional for a bit, it's possible some Simit changes caused the performance drop.

Even though this is the case, I think we should go ahead and merge this to make the GPU backend functional and remove the LLVM 3.4 dependence ASAP. I can continue investigating separately.

fredrikbk commented:
Ok, I agree. Continue investigating and I'll merge the change.

fredrikbk commented:
The LLVM folks just released 3.9 and this branch does not compile with it. We should fix that before we send out an email asking people to update their LLVM version.

gkanwar commented:
Thanks for fixing this up and merging! Looks like we will also need to point the master builders to LLVM 3.7.1. on the Jenkins server (similar to the PR builder).

fredrikbk commented:
Yes, could you do it? I have to set this machine up to be able to log in again at some point.

gkanwar commented:
Updated the builders, but looks like the Jenkins instance is out of hard drive space. I'll work on fixing this up tomorrow.
